pkg/yang/types.go

```diff
@@ -351,23 +351,47 @@ check:
 	// Patterns are ANDed according to section 9.4.6.  If all the patterns
 	// declared by t were also declared by the type t is based on, then
 	// no patterns are added.
-	patterns := map[string]bool{}
+	seenPatterns := map[string]bool{}
 	for _, p := range y.Pattern {
-		patterns[p] = true
+		seenPatterns[p] = true
 	}
-	for _, pv := range t.Pattern {
-		p := pv.Name
-		if _, err := syntax.Parse(p, syntax.Perl); err != nil {
+	seenPOSIXPatterns := map[string]bool{}
+	for _, p := range y.POSIXPattern {
+		seenPOSIXPatterns[p] = true
+	}
+
+	checkPattern := func(n Node, p string, flags syntax.Flags) {
+		_, err := syntax.Parse(p, flags)
+		if err != nil {
 			if re, ok := err.(*syntax.Error); ok {
 				// Error adds "error parsing regexp" to
 				// the error, re.Code is the real error.
 				err = errors.New(re.Code.String())
 			}
-			errs = append(errs, fmt.Errorf("%s: bad pattern: %v: %s", Source(pv), err, p))
+			errs = append(errs, fmt.Errorf("%s: bad pattern: %v: %s", Source(n), err, p))
 		}
-		if !patterns[p] {
-			patterns[p] = true
-			y.Pattern = append(y.Pattern, p)
+	}
+
+	// First parse out the pattern statements.
+	for _, pv := range t.Pattern {
+		checkPattern(pv, pv.Name, syntax.Perl)
```

Comment:
The arguments to the `pattern` statement aren't actually `syntax.Perl` - but this looks to be as close as we could get. Maybe just add a comment here that says something like "we know that these are actually W3C patterns, but Golang doesn't support those, so we rather treat them like Perl patterns". That being said -- maybe we just don't check these patterns, since right now -- we'd likely error on cases that are specifically those that are in the W3C standard that golang's `syntax.Perl` doesn't recognise.

I like the idea of checking them for the `POSIX` patterns, I'm just worried about the fidelity problem above.